test/tools/ossfuzz/strictasm_opt_ossfuzz.cpp

```diff
@@ -0,0 +1,36 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/interface/AssemblyStack.h>
+
+using namespace dev::solidity;
+using namespace std;
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
+{
+	if (_size > 450)
+		return 0;
+
+	string input(reinterpret_cast<char const*>(_data), _size);
+	AssemblyStack stack(EVMVersion::constantinople(), AssemblyStack::Language::StrictAssembly);
+
+	if (!stack.parseAndAnalyze("source", input))
+		return 0;
+
+	stack.optimize();
+	return 0;
```

Comment:
Do we want to test for codegen too? `.assemble()`

Reply:
I had this question too. @chriseth feels that fuzzing `.assemble()` will most likely trigger a (known) assertion for exceeding the number of stack slots that is currently supported (16 I think?)

Reply:
Ideally I'd run two versions of the fuzzer, one with optimization and one without. Without optimizations it definitely should run compile, with the optimizer on we can disable that for now.

Reply:
When you say "run compile", are you referring to the `compileEVM(adapter, false, _optimize);` API?

If so, I notice that this is called within `AssemblyStack::assemble(Machine _machine, bool _optimize)`. So, would the following harness be something that you have in mind for the other version (no optimization, just parse and assemble)?

```
	if (!stack.parseAndAnalyze("source", input))
		return 0;

	stack.assemble();
```

Reply:
When I say compile I mean calling `stack.assemble()`.

Reply:
We can compile, but then we should handle stack problems as not an error.

Reply:
@chriseth Would handling stack problems as not an error (i.e., ignoring them for the purpose of fuzz testing other code portions of the assembler) be as simple as removing the following assertion inside `EVMObjectCompiler::run`?

```
yulAssert(transform.stackErrors().empty(), "Stack errors present but not thrown.");
```

Reply:
No, this assertion should not be reachable. Please try it out with code like this one here: https://github.com/ethereum/solidity/blob/2c75c08a892574b3668589e4b76f5e914bbbeb96/test/libyul/CompilabilityChecker.cpp#L95

Reply:
@chriseth tested with the linked input, no assertions were raised.

Reply:
> Ideally I'd run two versions of the fuzzer, one with optimization and one without. Without optimizations it definitely should run compile

Updated the PR with two versions: one parses and optimizes, the other parses and assembles w/o optimization

Reply:
> No, this assertion should not be reachable. Please try it out with code like this one here:

Code like this triggers the `StackTooDeepError` exception
```
{function f(x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,x_11,x_12,x_13,x_14,x_15,x_16,x_17) {} }
```

I am curious why the test that you linked from `solidity/test/libyul/CompilabilityChecker.cpp` does not do this. My reasoning is that, in the latter case, the excessive stack slots are from inside the function scope (not parameter list) and the optimization removes them thereby saving stack slots.

On the other hand, the optimizer cannot optimize away function parameters.

Reply:
My guess is that if you run the full suite, it just removes the function, if it is not called anywhere.

Reply:
Update: I now catch `StackTooDeepError` to filter bug reports like above.